# ĐỒ ÁN CUỐI KỲ MÔN LẬP TRÌNH CHO KHOA HỌC DỮ LIỆU

*Giảng viên: Thầy Trần Trung Kiên*

---

**Nhóm sinh viên thực hiện:**

Sinh viên 01: Trương Trọng Lộc - 18120197

Sinh viên 02: Dương Nam Hải - 19120068

---

<h1>Nội Dung<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-các-thư-viện-cần-thiết" data-toc-modified-id="Import-các-thư-viện-cần-thiết-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import các thư viện cần thiết</a></span></li><li><span><a href="#Thu-thập-dữ-liệu" data-toc-modified-id="Thu-thập-dữ-liệu-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Thu thập dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Tổng-quan-về-dữ-liệu" data-toc-modified-id="Tổng-quan-về-dữ-liệu-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Tổng quan về dữ liệu</a></span></li><li><span><a href="#Dữ-liệu-được-phép-sử-dụng-hay-không?" data-toc-modified-id="Dữ-liệu-được-phép-sử-dụng-hay-không?-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Dữ liệu được phép sử dụng hay không?</a></span></li><li><span><a href="#Dữ-liệu-được-thu-thập-như-thế-nào?" data-toc-modified-id="Dữ-liệu-được-thu-thập-như-thế-nào?-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Dữ liệu được thu thập như thế nào?</a></span></li></ul></li><li><span><a href="#Khám-phá-dữ-liệu" data-toc-modified-id="Khám-phá-dữ-liệu-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Khám phá dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Đọc-dữ-liệu-từ-file-CSV-vào-dataframe" data-toc-modified-id="Đọc-dữ-liệu-từ-file-CSV-vào-dataframe-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Đọc dữ liệu từ file CSV vào dataframe</a></span></li><li><span><a href="#Dữ-liệu-gồm-có-bao-nhiêu-dòng-và-bao-nhiêu-cột?" data-toc-modified-id="Dữ-liệu-gồm-có-bao-nhiêu-dòng-và-bao-nhiêu-cột?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?</a></span></li><li><span><a href="#Mỗi-dòng-có-ý-nghĩa-gì?-Có-vấn-đề-các-dòng-có-ý-nghĩa-khác-nhau-không?" data-toc-modified-id="Mỗi-dòng-có-ý-nghĩa-gì?-Có-vấn-đề-các-dòng-có-ý-nghĩa-khác-nhau-không?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?</a></span></li><li><span><a href="#Dữ-liệu-có-các-dòng-bị-lặp-không?" data-toc-modified-id="Dữ-liệu-có-các-dòng-bị-lặp-không?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Dữ liệu có các dòng bị lặp không?</a></span></li><li><span><a href="#Mỗi-cột-có-ý-nghĩa-gì?" data-toc-modified-id="Mỗi-cột-có-ý-nghĩa-gì?-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Mỗi cột có ý nghĩa gì?</a></span></li><li><span><a href="#Mỗi-cột-hiện-đang-có-kiểu-dữ-liệu-gì?-Có-cột-nào-có-kiểu-dữ-liệu-chưa-phù-hợp-để-có-thể-xử-lý-tiếp-không?" data-toc-modified-id="Mỗi-cột-hiện-đang-có-kiểu-dữ-liệu-gì?-Có-cột-nào-có-kiểu-dữ-liệu-chưa-phù-hợp-để-có-thể-xử-lý-tiếp-không?-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?</a></span></li><li><span><a href="#Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-numeric,-các-giá-trị-được-phân-bố-như-thế-nào?" data-toc-modified-id="Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-numeric,-các-giá-trị-được-phân-bố-như-thế-nào?-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?</a></span></li><li><span><a href="#Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-categorical,-các-giá-trị-được-phân-bố-như-thế-nào?" data-toc-modified-id="Với-mỗi-cột-có-kiểu-dữ-liệu-dạng-categorical,-các-giá-trị-được-phân-bố-như-thế-nào?-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?</a></span></li></ul></li><li><span><a href="#Tiền-xử-lý" data-toc-modified-id="Tiền-xử-lý-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Tiền xử lý</a></span></li><li><span><a href="#Khám-phá-dữ-liệu-(tiếp-tục)" data-toc-modified-id="Khám-phá-dữ-liệu-(tiếp-tục)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Khám phá dữ liệu (tiếp tục)</a></span><ul class="toc-item"><li><span><a href="#Khảo-sát-mối-tương-quan-giữa-các-cột-dạng-numeric" data-toc-modified-id="Khảo-sát-mối-tương-quan-giữa-các-cột-dạng-numeric-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Khảo sát mối tương quan giữa các cột dạng numeric</a></span></li><li><span><a href="#Tiền-xử-lý" data-toc-modified-id="Tiền-xử-lý-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Tiền xử lý</a></span></li></ul></li><li><span><a href="#Khám-phá-dữ-liệu-(tiếp-tục)" data-toc-modified-id="Khám-phá-dữ-liệu-(tiếp-tục)-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Khám phá dữ liệu (tiếp tục)</a></span><ul class="toc-item"><li><span><a href="#Kiểm-tra-dữ-liệu-có-sự-bất-thường-ở-4-cột-country,-country_code,-region-và-year-không?" data-toc-modified-id="Kiểm-tra-dữ-liệu-có-sự-bất-thường-ở-4-cột-country,-country_code,-region-và-year-không?-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Kiểm tra dữ liệu có sự bất thường ở 4 cột country, country_code, region và year không?</a></span></li><li><span><a href="#Khảo-sát-về-tuổi-thọ-kỳ-vọng-ở-các-quốc-gia" data-toc-modified-id="Khảo-sát-về-tuổi-thọ-kỳ-vọng-ở-các-quốc-gia-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Khảo sát về tuổi thọ kỳ vọng ở các quốc gia</a></span></li><li><span><a href="#Khảo-sát-cột-Region" data-toc-modified-id="Khảo-sát-cột-Region-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Khảo sát cột Region</a></span><ul class="toc-item"><li><span><a href="#Có-tất-cả-bao-nhiêu-khu-vực-và-đó-là-những-khu-vực-nào?" data-toc-modified-id="Có-tất-cả-bao-nhiêu-khu-vực-và-đó-là-những-khu-vực-nào?-6.3.1"><span class="toc-item-num">6.3.1&nbsp;&nbsp;</span>Có tất cả bao nhiêu khu vực và đó là những khu vực nào?</a></span></li></ul></li></ul></li><li><span><a href="#Đưa-ra-câu-hỏi-cần-trả-lời" data-toc-modified-id="Đưa-ra-câu-hỏi-cần-trả-lời-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Đưa ra câu hỏi cần trả lời</a></span><ul class="toc-item"><li><span><a href="#Phân-tích-dữ-liệu-để-trả-lời-cho-các-câu-hỏi" data-toc-modified-id="Phân-tích-dữ-liệu-để-trả-lời-cho-các-câu-hỏi-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Phân tích dữ liệu để trả lời cho các câu hỏi</a></span><ul class="toc-item"><li><span><a href="#Câu-hỏi-1:-Các-quốc-gia-và-khu-vực-trên-thế-giới-có-tuổi-thọ-biến-động-như-thế-nào-từ-năm-2000-đến-năm-2016?-Việt-Nam-có-tuổi-thọ-như-thế-nào-so-với-các-quốc-gia-khác-trên-thế-giới-và-khu-vực?" data-toc-modified-id="Câu-hỏi-1:-Các-quốc-gia-và-khu-vực-trên-thế-giới-có-tuổi-thọ-biến-động-như-thế-nào-từ-năm-2000-đến-năm-2016?-Việt-Nam-có-tuổi-thọ-như-thế-nào-so-với-các-quốc-gia-khác-trên-thế-giới-và-khu-vực?-7.1.1"><span class="toc-item-num">7.1.1&nbsp;&nbsp;</span>Câu hỏi 1: Các quốc gia và khu vực trên thế giới có tuổi thọ biến động như thế nào từ năm 2000 đến năm 2016? Việt Nam có tuổi thọ như thế nào so với các quốc gia khác trên thế giới và khu vực?</a></span><ul class="toc-item"><li><span><a href="#Tổng-quan-thế-giới" data-toc-modified-id="Tổng-quan-thế-giới-7.1.1.1"><span class="toc-item-num">7.1.1.1&nbsp;&nbsp;</span>Tổng quan thế giới</a></span></li><li><span><a href="#Về-các-quốc-gia" data-toc-modified-id="Về-các-quốc-gia-7.1.1.2"><span class="toc-item-num">7.1.1.2&nbsp;&nbsp;</span>Về các quốc gia</a></span></li><li><span><a href="#Về-phía-từng-khu-vực" data-toc-modified-id="Về-phía-từng-khu-vực-7.1.1.3"><span class="toc-item-num">7.1.1.3&nbsp;&nbsp;</span>Về phía từng khu vực</a></span></li><li><span><a href="#Về-tuổi-thọ-của-Việt-Nam" data-toc-modified-id="Về-tuổi-thọ-của-Việt-Nam-7.1.1.4"><span class="toc-item-num">7.1.1.4&nbsp;&nbsp;</span>Về tuổi thọ của Việt Nam</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Nhìn-lại-quá-trình-làm-đồ-án" data-toc-modified-id="Nhìn-lại-quá-trình-làm-đồ-án-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Nhìn lại quá trình làm đồ án</a></span></li><li><span><a href="#Tài-liệu-tham-khảo" data-toc-modified-id="Tài-liệu-tham-khảo-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Tài liệu tham khảo</a></span></li></ul></div>

## Import các thư viện cần thiết

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

---

## Thu thập dữ liệu

### Tổng quan về dữ liệu

- Tập dữ liệu được sử dụng trong đồ án là tập dữ liệu về các **nhân tố có thể tác động đến tuổi thọ** con người 
(chẳng hạn về giáo dục, y tế, kinh tế) của 183 quốc gia trên thế giới từ năm 2000-2016. Tập dữ liệu tên là **WHO national life expectancy**.
- Tập dữ liệu được lấy và download từ Kaggle thông qua [đường dẫn](https://www.kaggle.com/mmattson/who-national-life-expectancy), bộ dữ liệu do tác giả MMattson thu thập.

### Dữ liệu được phép sử dụng hay không?

License của tập dữ liệu này là [CC0 - Public Domain](https://creativecommons.org/publicdomain/zero/1.0/). Với license này thì ta có thể sử dụng tập dữ liệu mà không cần xin phép tác giả. Cụ thể, license này thể hiện tác giả đã từ bỏ quyền sở hữu của bản thân để trao quyền cho người khác có thể sử dụng để chỉnh sửa, phân phối,... dưới bất kỳ hình thức nào trong phạm vi pháp luật (có thể là học tập, nghiên cứu, thương mại,...).

### Dữ liệu được thu thập như thế nào?

* Dữ liệu được thu thập từ dữ liệu của Tổ chức Y tế Thế giới (World Health Organization) thông qua cả 3 cách sau: 
    * Trang web dữ liệu Thống kê về Giáo dục các Quốc gia của Viện thống kê UNESCO ([UNESCO Databases of Resources on Education](https://en.unesco.org/themes/education/databases));
    * API từ [GHO (Global Health Observatory)](https://www.who.int/data/gho/);
    * Thủ công.
* Chi tiết về cách thu thập dữ liệu có thể tham khảo tại notebook của tác giả tại [đây](https://www.kaggle.com/mmattson/template-to-generate-the-dataset).

---

## Khám phá dữ liệu

### Đọc dữ liệu từ file CSV vào dataframe

Dữ liệu được lưu trong 1 file CSV duy nhất `who_life_exp.csv` và được chứa trong thư mục Data. Ta sử dụng thư viện Pandas để đọc dữ liệu từ CSV này vào Data Frame.

In [2]:
life_exp_df = pd.read_csv('./Data/who_life_exp.csv')
life_exp_df.head()

country country_code  region  year  life_expect  life_exp60  \
0  Angola          AGO  Africa  2000     47.33730    14.73400   
1  Angola          AGO  Africa  2001     48.19789    14.95963   
2  Angola          AGO  Africa  2002     49.42569    15.20010   
3  Angola          AGO  Africa  2003     50.50266    15.39144   
4  Angola          AGO  Africa  2004     51.52863    15.56860   

   adult_mortality  infant_mort  age1-4mort  alcohol  ...  che_gdp    une_pop  \
0         383.5583     0.137985    0.025695  1.47439  ...  1.90860  16395.473   
1         372.3876     0.133675    0.024500  1.94025  ...  4.48352  16945.753   
2         354.5147     0.128320    0.023260  2.07512  ...  3.32946  17519.417   
3         343.2169     0.122040    0.021925  2.20275  ...  3.54797  18121.479   
4         333.8711     0.115700    0.020545  2.41274  ...  3.96720  18758.145   

   une_infant  une_life  une_hiv  une_gni  une_poverty  une_edu_spend  \
0       122.2    46.522      1.0   2530.0         32.3        2.60753   
1       118.9    47.059      1.1   2630.0          NaN            NaN   
2       115.1    47.702      1.2   3180.0          NaN            NaN   
3       110.8    48.440      1.3   3260.0          NaN            NaN   
4       106.2    49.263      1.3   3560.0          NaN            NaN   

   une_literacy  une_school  
0           NaN         NaN  
1      67.40542         NaN  
2           NaN         NaN  
3           NaN         NaN  
4           NaN         NaN  

[5 rows x 32 columns]

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

In [3]:
num_rows, num_cols = life_exp_df.shape
num_rows, num_cols

(3111, 32)

**Nhận xét:** 
* Tập dữ liệu gồm 3111 dòng và 32 cột.
* Ngoài ra, theo thông tin của bộ dữ liệu, dữ liệu được khảo sát từ năm 2000-2016 ở 183 quốc gia, điều này hoàn toàn phù hợp với kết quả là 3111 dòng (nói cách dễ hiểu là do 2000-2016 là 17 năm, ta lấy 17 năm * 183 quốc gia = 3111 dòng, dự đoán ban đầu của nhóm em là vậy ^^ Nhóm em sẽ khảo sát ở phần sau xem thử có đúng như dự đoán không).

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

* Theo quan sát, mỗi dòng thể hiện thông tin về các yếu tố có thể ảnh hưởng đến tuổi thọ (tỉ lệ tiêm chủng, dân số, tỉ lệ người lớn biết chữ, thu nhập bình quân,...) của **một quốc gia** trong **một năm**. 
* Có vẻ **không** có dòng nào có ý nghĩa khác nhau. 

### Dữ liệu có các dòng bị lặp không?

Kiểm tra việc này và lưu kết quả vào biến `have_duplicated_rows`. Biến này sẽ có giá trị True nếu dữ liệu có các dòng bị lặp và có giá trị False nếu ngược lại.

In [4]:
have_duplicated_rows = life_exp_df.duplicated().sum()
if have_duplicated_rows != 0:
    print("Có dòng dữ liệu bị lặp")
else:
    print("Không có dữ liệu trùng lặp")

Không có dữ liệu trùng lặp


### Mỗi cột có ý nghĩa gì?

Theo mô tả trên Kaggle, mỗi cột có ý nghĩa như sau:

|Tên cột dữ liệu | Mô tả |
|:------:| ---|
|Country| Tên quốc gia khảo sát |
|country_code| Mã định danh quốc gia (3 chữ cái)|
|region| Khu vực của quốc gia khảo sát|
|year| Năm khảo sát|
|life_expect| Tuổi thọ kì vọng khi sinh (đơn vị năm) |
|life_exp60| Tuổi thọ kì vọng còn lại khi 60 tuổi (đơn vị năm)|
|adult_mortality|Tỷ lệ tử vong của người trưởng thành (xác suất tử vong từ 15 đến 60 tuổi trên 1000 dân số)|
|infant_mort|Tỷ lệ tử vong dưới 1 tuổi|
|age1-4mort|Tỷ lệ tử vong từ 1 đến 4 tuổi|
|alcohol|Mức tiêu thụ rượu bình quân đầu người từ 15 tuổi trở lên (tính bằng rượu nguyên chất)|
|bmi|BMI trung bình (từ 18 tuổi trở lên) (ước tính chuẩn hóa theo độ tuổi)|
|age5-19thinness|Tỷ lệ suy dinh dưỡng ở trẻ em và thanh thiếu niên (%)|
|age5-19obesity	|Tỷ lệ béo phì ở trẻ em và thanh thiếu niên (%)|
|hepatitis|Tỷ lệ tiêm ngừa viêm gan siêu vi B (HepB) ở trẻ 1 tuổi (%)|
|measles|Tỷ lệ tiêm ngừa sởi mũi đầu tiên (MCV1) ở trẻ 1 tuổi (%)|
|polio| Tỷ lệ tiêm ngừa bại liệt (Pol3) ở trẻ 1 tuổi (%)|
|diphtheria|Tỷ lệ tiêm ngừa bạch hầu, uốn ván và ho gà (DPT3) cho trẻ 1 tuổi (%)|
|basic_water|Tỷ lệ dân số được tiếp cận các dịch vụ nước cơ bản|
|doctors|Mật độ bác sĩ y khoa trên 10.000 dân|
|hospitals|Mật độ bệnh viện trên 100.000 dân|
|gni_capita|Tổng thu nhập quốc dân trên đầu người, đơn vị $ (tổng thu thập mà người dân và doanh nghiệp của quốc gia kiếm được)|
|gghe-d	|Chi tiêu từ ngân sách chính phủ cho lĩnh vực y tế tính theo tỷ lệ phần trăm của tổng sản phẩm quốc nội (GDP) (%)|
|che_gdp|Chi phí người dân bỏ ra cho dịch vụ y tế (CHE) tính theo phần trăm tổng sản phẩm quốc nội (GDP) (%)|
|une_pop|Dân số (đơn vị nghìn)|
|une_infant|Tỷ lệ tử vong ở trẻ sơ sinh trên 1.000 ca sinh sống)|
|une_life|Tuổi thọ kì vọng nếu các hình thức tử vong phổ biến tại thời điểm được sinh ra không thay đổi trong suốt cuộc đời của trẻ|
|une_hiv|Tỷ lệ nhiễm HIV, đơn vị \% trên tổng dân số từ 15-49 tuổi|
|une_gni|Thu nhập quốc dân trên đầu người|
|une_poverty|Tỷ lệ số người nghèo có mức sống 1,90 đô la một ngày (PPP) (\% trên dân số)|
|une_edu_spend|Chi tiêu của chính phủ cho giáo dục tính theo phần trăm GDP (\%)|
|une_literacy|Tỷ lệ người lớn biết chữ, cả hai giới (dân số trên 15 tuổi) (\%)|
|une_school|Số năm đi học trung bình (ISCED 1 trở lên), cả hai giới (dân số trên 25 tuổi)|

**Các thuộc tính có tên `une_`** được thống kê bởi tổ chức UNESCO.

**Các thuộc tính còn lại** được thống kê bởi GHO (Global Health Observatory).


### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [5]:
col_dtypes = life_exp_df.dtypes
col_dtypes

country             object
country_code        object
region              object
year                 int64
life_expect        float64
life_exp60         float64
adult_mortality    float64
infant_mort        float64
age1-4mort         float64
alcohol            float64
bmi                float64
age5-19thinness    float64
age5-19obesity     float64
hepatitis          float64
measles            float64
polio              float64
diphtheria         float64
basic_water        float64
doctors            float64
hospitals          float64
gni_capita         float64
gghe-d             float64
che_gdp            float64
une_pop            float64
une_infant         float64
une_life           float64
une_hiv            float64
une_gni            float64
une_poverty        float64
une_edu_spend      float64
une_literacy       float64
une_school         float64
dtype: object

**Nhận xét:** Các cột có kiểu dữ liệu không phải `object` phù hợp, cần làm rõ hơn 3 cột có kiểu dữ liệu đang là `object`, trong Python kiểu dữ liệu `object` thường chỉ chuỗi, nhưng thật chất nó có thể chứa đối tượng bất kỳ vì bên dưới lưu dưới dạng địa chỉ. Trong Data Frame, nếu một cột có `dtype` là `object` thì có thể các phần tử trong cột này sẽ có kiểu dữ liệu khác nhau. Do đó, cần khảo sát từng cột có `dtype` là `object`.

In [6]:
# Hàm trả về các kiểu dữ liệu trong 1 cột
def open_object_dtype(s):
    dtypes = set()
    
    dtypes = set(s.apply(type))
    
    return dtypes

In [7]:
print(open_object_dtype(life_exp_df['country']))
print(open_object_dtype(life_exp_df['country_code']))
print(open_object_dtype(life_exp_df['region']))

{<class 'str'>}
{<class 'str'>}
{<class 'str'>}


Như vậy, cả 3 cột đều có 1 kiểu dữ liệu duy nhất là `str`. 
Do đó **tất cả các cột** trong tập dữ liệu đều đang có kiểu dữ liệu **phù hợp**.

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

* Trong 32 cột dữ liệu của dataset, theo khám phá ở trên có 3 cột là `country`, `country_code` và `region` dạng `object` (cụ thể hơn là `str`). Do đó, ta có tổng cộng 29 cột có kiểu dữ liệu dạng numeric (bài này là `in64` và `float64`). 
* Với mỗi cột numeric, tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max. Kết quả được vào Data Frame `nume_col_df`. Dataframe này có 9 dòng là "missing_ratio", "count", "mean", "std", "min", "25%", "50%", "75%", "max"; và có 29 cột tương ứng với các cột có `dtype` là `int64` hoặc `float64`.  

In [8]:
def missing_ratio(s):
    return s.isna().mean() * 100
missing_df = life_exp_df.iloc[:,3:].agg([missing_ratio])
nume_col_df = pd.concat([missing_df, life_exp_df.iloc[:,3:].describe()])
nume_col_df

year  life_expect   life_exp60  adult_mortality  \
missing_ratio     0.000000     0.000000     0.000000         0.000000   
count          3111.000000  3111.000000  3111.000000      3111.000000   
mean           2008.000000    69.146384    18.912060       193.527786   
std               4.899767     9.129761     2.885412       114.007447   
min            2000.000000    36.227360    10.728000        49.201850   
25%            2004.000000    63.200095    16.619485       108.269750   
50%            2008.000000    71.597200    18.508260       164.822200   
75%            2012.000000    75.537030    21.099175       250.832300   
max            2016.000000    84.166160    26.394020       696.874200   

               infant_mort   age1-4mort      alcohol          bmi  \
missing_ratio     0.000000     0.000000     1.607200     1.092896   
count          3111.000000  3111.000000  3061.000000  3077.000000   
mean              0.032496     0.003489     4.834781    25.052714   
std               0.030361     0.005146     3.998741     2.193445   
min               0.001470     0.000065     0.000000    19.800000   
25%               0.008255     0.000355     1.198030    23.300000   
50%               0.019995     0.000895     3.993560    25.500000   
75%               0.051720     0.004877     7.723420    26.500000   
max               0.164515     0.039095    20.182460    32.200000   

               age5-19thinness  age5-19obesity  ...      che_gdp  \
missing_ratio         1.092896        1.092896  ...     3.760849   
count              3077.000000     3077.000000  ...  2994.000000   
mean                  5.312447        5.972278  ...     6.110353   
std                   4.652408        4.530812  ...     2.505267   
min                   0.100000        0.100000  ...     1.025160   
25%                   1.800000        2.000000  ...     4.238798   
50%                   3.800000        5.200000  ...     5.758030   
75%                   7.800000        8.900000  ...     7.850327   
max                  28.100000       26.700000  ...    20.413410   

                    une_pop   une_infant     une_life      une_hiv  \
missing_ratio  1.189328e+00     0.000000     0.000000    23.818708   
count          3.074000e+03  3111.000000  3111.000000  2370.000000   
mean           3.707550e+04    30.488428    68.957352     2.038397   
std            1.378377e+05    27.756930     9.359194     4.593804   
min            7.601600e+01     1.600000    39.441000     0.100000   
25%            2.195105e+03     8.000000    62.844500     0.100000   
50%            8.544297e+03    19.500000    71.409000     0.400000   
75%            2.509552e+04    48.050000    75.570645     1.500000   
max            1.414049e+06   142.400000    83.984880    28.200000   

                     une_gni  une_poverty  une_edu_spend  une_literacy  \
missing_ratio       3.760849    70.652523      41.337191     81.645773   
count            2994.000000   913.000000    1825.000000    571.000000   
mean            14964.832999    10.845893       4.532930     81.984472   
std             17495.137508    17.168798       1.754280     19.665588   
min               420.000000     0.100000       0.787440     14.376040   
25%              2970.000000     0.600000       3.262830     72.701130   
50%              8340.000000     3.100000       4.425410     90.953740   
75%             20482.500000    12.400000       5.494980     95.786975   
max            122670.000000    94.100000      14.059080     99.998190   

               une_school  
missing_ratio   74.124076  
count          805.000000  
mean             9.712214  
std              2.854723  
min              0.559340  
25%              7.735940  
50%             10.270400  
75%             12.070610  
max             14.378800  

[9 rows x 29 columns]

**Nhận xét:** Sau khi khám phá sự phân bố của các cột dữ liệu, ta nhận thấy có 4 cột numeric với missing ratio trên 50% là "hospitals", "une_poverty", "une_literacy", "une_school". Việc dùng các cột nào trong phân tích dữ liệu có thể cho ra kết quả không tốt, vì vậy ta cần loại bỏ chúng trong bước tiền xử lý.

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

* Theo khám phá ở trên có 3 cột là `country`, `country_code` và `region` dạng `object` (cụ thể hơn là `str`). Do đó, kiểu dữ liệu dạng categorical của tập dữ liệu này gồm 3 cột đẫ nêu. 
* Với mỗi cột dạng categorical, tính tỉ lệ % giá trị thiếu (từ 0 đến 100), số lượng giá trị khác nhau (không xét giá trị thiếu), list/array các giá trị khác nhau (không xét giá trị thiếu). Kết quả được lưu vào Data Frame `cate_col_df`. Dataframe này có 3 dòng là "missing_ratio", "num_diff_vals", "diff_vals"; và có 3 cột là 'country', 'country_code', 'region'. 

In [9]:
def unique_not_miss(s):
    return s.dropna().unique()
cate_col_df = life_exp_df.iloc[:,:3] \
                            .agg([missing_ratio, pd.Series.nunique, unique_not_miss])

cate_col_df.index = ['missing_ratio','num_diff_vals','diff_vals']
cate_col_df

country  \
missing_ratio                                                0.0   
num_diff_vals                                                183   
diff_vals      [Angola, Burundi, Benin, Burkina Faso, Botswan...   

                                                    country_code  \
missing_ratio                                                0.0   
num_diff_vals                                                183   
diff_vals      [AGO, BDI, BEN, BFA, BWA, CAF, TZA, UGA, ZAF, ...   

                                                          region  
missing_ratio                                                0.0  
num_diff_vals                                                  6  
diff_vals      [Africa, Americas, Eastern Mediterranean, Euro...

**Nhận xét:** Cả 3 cột dạng categorical đều không có giá trị thiếu. Hai cột `country` và `country_code` có 183 giá trị khác nhau, riêng cột `region` có 6 giá trị khác nhau. Điều này hoàn toàn phù hợp vì có 183 quốc gia được khảo sát trong tập dữ liệu này, các quốc gia khác nhau có thể thuộc chung 1 khu vực.

## Tiền xử lý

Ta sẽ thực hiện drop các cột với tỉ lệ thiếu hụt thông tin lớn hơn 50%, sau đó lưu vào 1 Data Frame mới là `life_exp_pro_df` để sử dụng cho việc phân tích về sau.

In [10]:
missing_df = life_exp_df.agg(missing_ratio)
cols_to_drop = missing_df[missing_df > 50]
cols_to_drop

hospitals       95.821279
une_poverty     70.652523
une_literacy    81.645773
une_school      74.124076
dtype: float64

In [11]:
life_exp_pro_df = life_exp_df.drop(cols_to_drop.index, axis = 1)
life_exp_pro_df.shape

(3111, 28)

Sau khi tiền xử lý xóa 4 cột, bộ dữ liệu còn 3111 dòng và 28 cột dữ liệu. Ta sẽ tiến hành phân tích trên dataframe `life_exp_pro_df` này.

## Khám phá dữ liệu (tiếp tục)

### Khảo sát mối tương quan giữa các cột dạng numeric

Các cột của tập dữ liệu thể hiện các nhân tố ảnh hưởng đến tuổi thọ, vì thế ta cần khảo sát mối tương quan giữa các cột này (dạng numeric). Để làm được, ta thực hiện: 
* Tính độ tương quan giữa các cột bằng phương thức `corr` của Pandas.
* Trực quan hóa ma trận thể hiện độ tương quan đã tính thông qua thư viện Matplotlib.

**Bước 01:** Tính độ tương quan giữa các cột


In [12]:
corr = life_exp_pro_df.corr()
corr

year  life_expect  life_exp60  adult_mortality  \
year             1.000000     0.190961    0.192872        -0.170946   
life_expect      0.190961     1.000000    0.886159        -0.946419   
life_exp60       0.192872     0.886159    1.000000        -0.775321   
adult_mortality -0.170946    -0.946419   -0.775321         1.000000   
infant_mort     -0.195293    -0.930113   -0.769839         0.813210   
age1-4mort      -0.191723    -0.864414   -0.661082         0.756230   
alcohol          0.006428     0.399797    0.465785        -0.246530   
bmi              0.170244     0.597996    0.458839        -0.517095   
age5-19thinness -0.081810    -0.565396   -0.556536         0.455205   
age5-19obesity   0.278359     0.621000    0.547764        -0.571792   
hepatitis        0.128042     0.341137    0.238058        -0.276921   
measles          0.142717     0.640408    0.504875        -0.526185   
polio            0.128699     0.646486    0.501654        -0.537973   
diphtheria       0.141351     0.640473    0.503303        -0.529967   
basic_water      0.109352     0.831099    0.680990        -0.732704   
doctors          0.012124     0.651891    0.571336        -0.547177   
gni_capita       0.118967     0.585603    0.582539        -0.527650   
gghe-d           0.081852     0.609787    0.670995        -0.485836   
che_gdp          0.112887     0.284920    0.388805        -0.198618   
une_pop          0.016757     0.032126    0.004131        -0.061503   
une_infant      -0.187952    -0.931899   -0.776510         0.814884   
une_life         0.185664     0.991448    0.869439        -0.943640   
une_hiv         -0.018294    -0.511821   -0.343894         0.680558   
une_gni          0.147887     0.603019    0.602258        -0.544664   
une_edu_spend    0.021240     0.261548    0.280849        -0.147110   

                 infant_mort  age1-4mort   alcohol       bmi  age5-19thinness  \
year               -0.195293   -0.191723  0.006428  0.170244        -0.081810   
life_expect        -0.930113   -0.864414  0.399797  0.597996        -0.565396   
life_exp60         -0.769839   -0.661082  0.465785  0.458839        -0.556536   
adult_mortality     0.813210    0.756230 -0.246530 -0.517095         0.455205   
infant_mort         1.000000    0.909032 -0.444580 -0.641874         0.578230   
age1-4mort          0.909032    1.000000 -0.305744 -0.615242         0.486846   
alcohol            -0.444580   -0.305744  1.000000  0.272832        -0.481989   
bmi                -0.641874   -0.615242  0.272832  1.000000        -0.685819   
age5-19thinness     0.578230    0.486846 -0.481989 -0.685819         1.000000   
age5-19obesity     -0.602402   -0.544587  0.140611  0.807143        -0.548601   
hepatitis          -0.425655   -0.388905  0.119041  0.291555        -0.265812   
measles            -0.720169   -0.698573  0.291679  0.468788        -0.439146   
polio              -0.728603   -0.700788  0.289802  0.441274        -0.415575   
diphtheria         -0.720741   -0.694145  0.294311  0.446225        -0.418306   
basic_water        -0.851514   -0.802324  0.407690  0.673261        -0.474768   
doctors            -0.644034   -0.569178  0.560723  0.463434        -0.582657   
gni_capita         -0.539514   -0.417762  0.304989  0.410808        -0.370173   
gghe-d             -0.576804   -0.447373  0.529580  0.461679        -0.573455   
che_gdp            -0.238393   -0.162339  0.372205  0.248024        -0.396721   
une_pop            -0.003967   -0.032953 -0.037812 -0.166194         0.265009   
une_infant          0.996975    0.902324 -0.449387 -0.644441         0.586277   
une_life           -0.926927   -0.855848  0.394807  0.587912        -0.551026   
une_hiv             0.341615    0.292005 -0.038341 -0.154689         0.193470   
une_gni            -0.549928   -0.421805  0.296187  0.413644        -0.368289   
une_edu_spend      -0.318738   -0.283001  0.215987  0.312443        -0.281627   

                 age5-19obesity  ...   doctors  gni_capita    gghe-

**Bước 02:** Trực quan hóa ma trận thể hiện độ tương quan thông qua heatmap của thư viện Seaborn.


In [ ]:
mask = np.zeros_like(corr, dtype=bool)
mask[np.triu_indices_from(mask)] = True
plt.subplots(figsize=(18,12))
dataplot = sns.heatmap(corr, cmap="coolwarm", annot=True, mask=mask)
plt.title("Correlation Heatmap")
plt.show()

**Nhận xét:** Từ Heatmap trên, ta rút ra:
* Các cặp (`une_life`, `life_expect`), (`une_infant`, `infant_mort`) và (`une_gni`, `gni_capita`) có độ tương quan cao thể hiện sự đồng biến (0.99 và 1.00). 
* Các cặp (`adult_mortality`, `life_expect`), (`une_life`, `adult_mortality`), (`une_life`, `infant_mort`),... có độ tương quan thấp (nghịch biến).
* Các mối quan hệ giữa 2 cột có độ tương quan càng cao hoặc càng thấp (nghĩa là càng tiến về 1 nếu đồng biến hoặc -1 nếu nghịch biến) càng có sự ảnh hưởng lẫn nhau và được xem là "good relationship". Các mối quan hệ có độ tương quan càng tiến về 0 cho thấy sự thay đổi giữa 2 cột không ổn định (nghĩa là khi cột này tăng, cột kia có thể tăng hoặc giảm, không có sự ổn định nhất định).

### Tiền xử lý

Quan sát các cặp cột có độ tương quan là 1.00 ở nhận xét trên, kết hợp với việc khảo sát sự phân bổ các giá trị numeric ở trên, ta thấy rằng: 
* Với 2 cột `une_infant` và `infant_mort`, tỉ lệ thiếu cả 2 cột là 0%, các giá trị về min, max, mean đều xấp xỉ nhau, chỉ khác về đơn vị (`une_infant` được tính trên 1000 người). 
* Với 2 cột `une_gni`, `gni_capita`, cột `gni_capita` có tỉ lệ thiếu là 21.92% trong khi đó cột `une_gni` tỉ lệ thiếu là 3.76%. 
Do đó, ta tiến hành loại bỏ các cột `infant_mort` và `gni_capita`

In [ ]:
life_exp_pro_df = life_exp_pro_df.drop(columns=['infant_mort','gni_capita'])
life_exp_pro_df.shape

## Khám phá dữ liệu (tiếp tục)

### Kiểm tra dữ liệu có sự bất thường ở 4 cột country, country_code, region và year không? 

Mỗi nước phải được khảo sát trong 17 năm, liên tục từ 2000 đến 2016, country_code và region phải giống nhau qua từng năm ở mỗi nước qua trong các năm khảo sát. Kết quả được lưu vào biến `is_normal`, nếu biến này là `True` thì dữ liệu bình thường, hợp lý; ngược lại biến này mang giá trị là `False`.

In [ ]:
values, counts = np.unique(life_exp_pro_df.groupby(['country', 'country_code', 'region']).size().values, \
                           return_counts=True)

max_year = life_exp_pro_df['year'].max()
min_year = life_exp_pro_df['year'].min()

num_countries = life_exp_pro_df['country'].nunique()

if len(values) == 1 and len(counts) == 1 \
    and values[0] == max_year - min_year + 1 and counts[0] == num_countries:
    is_normal = True
else: 
    is_normal = False
    
is_normal

### Khảo sát về tuổi thọ kỳ vọng ở các quốc gia

In [ ]:
fig, ax = plt.subplots(figsize=(16, 9))

cdict = {'Africa': 'orange', 'Americas': 'red', 'Western Pacific': 'blue', \
         'South-East Asia': 'green', 'Europe': 'purple', 'Eastern Mediterranean':'yellow'}

grouped = life_exp_pro_df.groupby('region')
for key, group in grouped:
    group.plot(ax=ax, kind='scatter', x='year', y='life_expect', label=key, color=cdict[key])
plt.show()

**Nhận xét:** Ta thấy dữ liệu có 1 điểm bất thường, điểm này thể hiện ở khu vực châu Mỹ vào năm 2010. Ta tiếp tục khảo sát tỉ lệ tử vong xem có gì bất thường không?

Để có thể lấy được tỉ lệ tử vong, ta cần tính tổng 3 cột thể hiện tỉ lệ tử vong vốn có là `une_infant` (Tỉ lệ tử vong trẻ sơ sinh), `age1-4mort` (Tỉ lệ tử vong từ 1-4 tuổi) và `adult_mortality` (Tỉ lệ tử vong ở người trưởng thành).

Tuy nhiên để làm được việc này ta cũng cần biến đổi cột `age1-4mort` từ đơn vị tỉ lệ trên tổng số dân thành trên 1000 người giống 2 cột còn lại. Ta sẽ nhân cột `age1-4mort` cho 1000.

In [ ]:
fig, ax = plt.subplots(figsize=(16, 9))

mort_df = life_exp_pro_df[['region', 
                           'year',
                           'country']]
mort_df.insert(3, "mort_rate", life_exp_pro_df['une_infant'] \
                        + life_exp_pro_df['age1-4mort'] * 1000 \
                        + life_exp_pro_df['adult_mortality'])
# mort_df.['mort_rate'] = ...
mean_mort = mort_df.groupby('year')['mort_rate'].mean()
mean_mort.plot.line()
grouped = mort_df.groupby('region')

cdict = {'Africa': 'orange', 'Americas': 'red', 'Western Pacific': 'blue', \
         'South-East Asia': 'green', 'Europe': 'purple', 'Eastern Mediterranean':'yellow'}

for key, group in grouped:
    group.plot(ax=ax, kind='scatter', x='year', y='mort_rate', label=key, color=cdict[key])
    
plt.title('The distribution of life expectancy by country by year',size=16)
plt.show()

**Nhận xét:** Tỉ lệ tử vong của 1 quốc gia nào đó ở chây Mỹ năm 2010 tăng bất thường. Ta kiểm tra xem đây là quốc gia nào.

In [ ]:
ame_2010_df = mort_df[(mort_df['region'] == 'Americas') 
                          & (mort_df['year'] == 2010)]
idx_max = ame_2010_df['mort_rate'].idxmax()
max_mort_country = ame_2010_df.loc[idx_max]['country']
mort_df[mort_df['country'] == max_mort_country]

Vậy quốc gia gây ra outlier chính là Haiti, với tỉ lệ tử vong năm 2010 vào xấp xỉ 789/1000 người, tăng khoảng gấp đôi so với các năm còn lại. Khám phá thêm về Haiti năm 2010 ta tìm được thông tin về [trận động đất năm 2010](https://www.history.com/this-day-in-history/massive-earthquake-strikes-haiti) gây thương vong vào khoản 200,000 người. Đó là nguyên nhân gây ra outlier trong tập dữ liệu.

Ta nhận thấy có thể dùng dataframe `mort_df` đã có để khảo sát thêm liệu còn các trường hợp quốc gia nào có biến động tương tự hay không. Ta sẽ vẽ line graph cho từng quốc gia trong từng khu vực qua các năm.

In [ ]:
region_names = mort_df['region'].unique()
fig, axs = plt.subplots(3, 2)
fig.set_size_inches(12.0, 20.0)
for ireg, region in enumerate(region_names):
    ix = ireg//2
    iy = ireg%2
    axs[ix, iy].set_title(region)
    temp_df = mort_df[mort_df['region'] == region]
    for country in temp_df['country'].unique():
        axs[ix, iy].plot(temp_df[temp_df['country']==country]['year'], 
                         temp_df[temp_df['country']==country]['mort_rate'])
    axs[ix, iy].set_xlabel('year')
    axs[ix, iy].set_ylabel('mortality rate')
plt.tight_layout()
plt.show()

Ngoài khu vực Americas ra, khu vực SEA có một quốc gia cũng tăng đột tiến tỉ lệ tử vong vào năm 2008. Ta sẽ xem đây là quốc gia nào.

In [ ]:
sea_2008_df = mort_df[(mort_df['region'] == 'South-East Asia') 
                          & (mort_df['year'] == 2008)]
idx_max = sea_2008_df['mort_rate'].idxmax()
max_mort_country = sea_2008_df.loc[idx_max]['country']
mort_df[mort_df['country'] == max_mort_country]

Vậy quốc gia đó là Myanmar. Sau khi nhóm em thử tra cứu về Myanmar năm 2008 thì tìm hiểu được vào ngày 2/5/2008 có [cơn bão Nargis] (https://en.wikipedia.org/wiki/Cyclone_Nargis) diễn ra khiến ít nhất 138,373 người thiệt mạng. Như vậy xảy ra outlier này là hoàn toàn hợp lý.

Ngoài ra thấy ở khu vực Eastern Mediterranean cũng có một quốc gia có tỉ lệ tử vong tăng liên tục từ năm 2011 đến năm 2014. Ta khảo sát thử xem đây là quốc gia nào.

In [ ]:
 mort_dif_df = mort_df[(mort_df['region'] == 'Eastern Mediterranean') 
                          & ((mort_df['year'] == 2011) \
                             | (mort_df['year'] == 2014))]
idx_max = pd.DataFrame(abs(mort_dif_df['mort_rate'].values[1::2] \
                           - mort_dif_df['mort_rate'].values[::2]), 
             columns=['mort_rate']).idxmax()
EM_country = mort_dif_df.reset_index().loc[idx_max * 2]['country'].values[0]
mort_df[mort_df['country'] == EM_country]

Vậy quốc gia đó là Ả Rập Syrian. Ta sẽ  Sau khi tìm hiểu về Syrian năm 2011, nhóm đã phát hiện từ tháng 3 năm 2011 Syrian xảy ra [cuộc nội chiến đa đảng](https://en.wikipedia.org/wiki/Human_rights_in_Syria) và kéo dài đến tận ngày nay, gây thương vong về người rất lớn. Theo số liệu, tỉ lệ tử vong tăng mạnh (khoảng gấp đôi) từ 2011 đến 2014 và vẫn còn cao trong những năm sau đó.

- Ta thấy cần loại bỏ outlier của quốc gia Haiti vì mức tăng là khoảng 200% gây ra để tránh kết quả xấu khi thực hiện phân tích dữ liệu. 
- Trái lại, đối với trường hợp Myanmar, outlier này có mức tăng không đáng kể như Haiti. Ta không cần loại outlier này.
- Đối với Ả Rập Syrian, tuy mức tăng cũng vào khoảng trên 200% (năm 2014 so với 2011) nhưng những năm sau đó tỉ lệ tử vong ổn định và không có đột biến. Ta cũng giữ lại không loại outlier này.

=> Ta sẽ tiến hành loại dòng dữ liệu Haiti năm 2010

In [ ]:
life_exp_pro_df.drop(life_exp_pro_df[(life_exp_pro_df['country'] == 'Haiti') 
                             & (life_exp_pro_df['year'] == 2010)].index, 
                 inplace=True)
life_exp_pro_df.shape

### Khảo sát cột Region

Quan sát một vài giá trị đầu và cuối của dữ liệu

In [ ]:
life_exp_pro_df

**Nhận xét:** Nước Lào thuộc vào khu vực Tây Thái Bình Dương (đáng lẽ theo lẽ thường Lào phải thuộc về Đông Nam Á (South-East Asia)). Ta kiểm tra kỹ hơn về các khu vực có trong tập dữ liệu.

#### Có tất cả bao nhiêu khu vực và đó là những khu vực nào?

In [ ]:
life_exp_pro_df.groupby(['region']).size() / (max_year - min_year + 1)

Có 6 khu vực trong tập dữ liệu, khu vực châu Phi có 47 nước, châu Mỹ 33 nước, Đông Địa Trung Hải 21 nước, châu Âu 50 nước, Đông Nam Á 11 nước, Tây Thái Bình Dương 21 nước. Tiếp túc kiểm tra 2 khu vực Tây Thái Bình Dương và Đông Nam Á, xem thử có nước nào như nước Lào không?

In [ ]:
life_exp_pro_df[life_exp_pro_df['region'] == 'Western Pacific']['country'].unique()

In [ ]:
life_exp_pro_df[life_exp_pro_df['region'] == 'South-East Asia']['country'].unique()

**Nhận xét:** Dường như cột `region` không phân chia theo khu vực mà đã học trong môn địa lý. Theo nhóm em có tìm hiểu thì WHO sẽ chia các nước trên thế giới thành 6 khu vực, ở mỗi khu vực sẽ có các văn phòng tương ứng. Cụ thể có thể xem tại [đây](https://www.who.int/about/who-we-are/regional-offices) hoặc [Wikipedia](https://en.wikipedia.org/wiki/List_of_WHO_regions).

---

## Đưa ra câu hỏi cần trả lời

* **Câu hỏi 1:** Các quốc gia và khu vực trên thế giới có tuổi thọ biến động như thế nào từ năm 2000 đến năm 2016? Việt Nam có tuổi thọ như thế nào so với các quốc gia khác trên thế giới và khu vực?
	
* **Câu hỏi 2:** Các vấn đề về y tế, kinh tế, xã hội có tác động như thế nào đến tuổi thọ của các quốc gia và các khu vực trên thế giới trong những năm 2000 đến năm 2016?

	* Y tế: Khảo sát sự tác động của tỷ lệ tiêm chủng (viêm gan, sởi, bại liệt, bạch hầu). 
	* Kinh tế: Khảo sát sự tác động của tổng thu nhập quốc dân, chi tiêu của chính phủ cho y tế, chi phí dân bỏ ra cho y tế, chi tiêu cho giáo dục giới tính.
	* Xã hội: Khảo sát sự tác động của tiêu thụ rượu, tỷ lệ người dân tiếp cận được nước sạch, tỷ lệ HIV/AIDS, dân số.

* **Câu hỏi 3:** Các vấn đề về y tế, kinh tế, xã hội ở các quốc gia đang phát triển và các quốc gia phát triển có gì khác biệt không? Nếu có thì sự khác biệt này như thế nào?

### Phân tích dữ liệu để trả lời cho các câu hỏi 

#### Câu hỏi 1: Các quốc gia và khu vực trên thế giới có tuổi thọ biến động như thế nào từ năm 2000 đến năm 2016? Việt Nam có tuổi thọ như thế nào so với các quốc gia khác trên thế giới và khu vực?

**Ý nghĩa khi trả lời câu hỏi:** Biết được cái nhìn tổng quan về tuổi thọ trung bình của các quốc gia (trong đó có Việt Nam) cũng như các khu vực trên thế giới, cụ thể: 
* Về phía Chính phủ, Nhà nước: Biết được quốc gia mình đang ở vị trí nào trong khu vực và thế giới, từ đó đề ra các giải pháp, chiến lược để duy trì và phát huy nhằm đạt được tuổi thọ trung bình cao hơn.
* Về phía người dân: có thể là **một trong số** các nhân tố dùng để tham khảo khi có ý định du học, định cư,...
* Ngoài ra, tạo tiền đề để các phân tích, nghiên cứu khác ra đời.

**Hướng tiếp cận để trả lời câu hỏi:**
Để trả lời câu hỏi này, nhóm tiếp cận theo hướng tách nhỏ thành các câu hỏi nhỏ hơn, cụ thể:
* Tổng quan thế giới:
    * Tuổi thọ trung bình trên thế giới biến động như thế nào qua các năm?
* Về phía từng quốc gia: 
    * Tuổi thọ trung bình của các quốc gia như thế nào?
    * Top 10 quốc gia có tuổi thọ trung bình cao nhất, thấp nhất?
    * Độ chênh lệch tuổi thọ năm 2016 và năm 2000 của các quốc gia như thế nào? 
    * Giải thích sự biến động về mặt tuổi thọ trung bình của các quốc gia trên thế giới.
* Về phía từng khu vực:
    * Tuổi thọ trung bình của các khu vực như thế nào? Đâu là khu vực có tuổi thọ trung bình cao nhất, thấp nhất?
    * Nhận xét về sự phân bố tuổi thọ trung bình của các khu vực trên thế giới.
* Về tuổi thọ của Việt Nam:
    * Việt Nam đang ở vị trí nào về mặt tuổi thọ trên thế giới và trong khu vực? 
    * Tốc độ biến động tuổi thọ ở Việt Nam qua các năm có xu hướng như thế nào?

##### Tổng quan thế giới

**Tuổi thọ trung bình trên thế giới biến động như thế nào qua các năm?**

Các bước thực hiện: 
* **Bước 1:** Tính trung bình tuổi thọ của 183 quốc gia trong tập dữ liệu theo từng năm. 
* **Bước 2:** Vẽ đồ thị thể hiện sự biến động tuổi thọ trung bình của các quốc gia trên thế giới. 
* **Bước 3:** Vẽ đồ thị histogram thể hiện phân phối tuổi thọ trung bình của thế giới năm 2000 và 2016.
* **Bước 4:** Kết luận.

**Bước 1:** Tính trung bình tuổi thọ của 183 quốc gia trong tập dữ liệu theo từng năm. Kết quả được lưu vào `life_exp_mean`

Tiến hành gom nhóm dữ liệu theo năm (`year`). Sau đó, tính trung bình trên cột `life_expect`. Kết quả thu được lưu vào Series `life_exp_mean`.

In [ ]:
life_exp_mean = life_exp_pro_df.groupby('year').agg(pd.Series.mean)['life_expect']
life_exp_mean

**Bước 2:** Vẽ đồ thị đường thể hiện sự biến động tuổi thọ trung bình của các quốc gia trên thế giới

In [ ]:
fig = plt.figure(figsize = (10, 5))
 
life_exp_mean.plot.line(ylabel='Life Expectancy (years)',xlabel='Year', marker='o')
 
plt.title("Mean life expectancy of countries in the world (2000-2016)", weight='bold', fontsize=14)
plt.tight_layout()
plt.show()

**Nhận xét:** Tuổi thọ trung bình của thế giới tăng qua các năm, cụ thể năm 2016 tăng so với năm 2000 khoảng 5 tuổi (~8.14%). Nguyên nhân có lẽ là do thế giới ngày càng phát triển và hiện đại, sức khỏe được quan tâm hơn trước dẫn đến tuổi thọ tăng.

**Bước 3:** Vẽ đồ thị histogram thể hiện phân phối tuổi thọ trung bình của thế giới năm 2000 và 2016

Để phân tích rõ hơn sự biến động tuổi thọ ở năm 2000 và năm 2016, tiến hành vẽ biểu đồ histogram và đường biểu diễn mật độ số lượng quốc gia thể hiện sự phân phối tuổi thọ trung bình. Nhóm sử dụng thư viện Seaborn để hỗ trợ vẽ kết hợp biểu đồ histogram và đường mật độ thay vì thư viện Matplotlib.

In [ ]:
fig = plt.figure(figsize = (10, 10))

sns.histplot(life_exp_pro_df[life_exp_pro_df.year.eq(life_exp_pro_df.year.min())]["life_expect"],
             color='orange', label=str(life_exp_pro_df.year.min()),
             kde=True, stat="density", linewidth=0)
sns.histplot(life_exp_pro_df[life_exp_pro_df.year.eq(life_exp_pro_df.year.max())]["life_expect"],
             color='blue', label=str(life_exp_pro_df.year.max()),
             kde=True, stat="density", linewidth=0)
plt.legend()
plt.xlabel("Life Expectancy (years)")
plt.ylabel("Density of countries")
plt.title("Life Expectancy Distribution",
          fontsize=14, weight='bold')
plt.tight_layout()
plt.show()

**Nhận xét:** Ta thấy sự khác biệt về tuổi thọ giữa các quốc gia có xu hướng giảm xuống và có sự gia tăng tuổi thọ từ năm 2000 đến năm 2016.

**Bước 4:**  Kết luận

Nhìn chung về tổng quan thế giới, xu hướng tuổi thọ trung bình có sự gia **tăng** qua từng năm. 

##### Về các quốc gia

**Tuổi thọ trung bình của các quốc gia như thế nào?**

Các bước thực hiện: 
* Bước 1: Tính trung bình tuổi thọ của **từng** quốc gia trong tập dữ liệu trong 17 năm từ năm 2000 đến 2016.
* Bước 2: Trực quan hóa giá trị đã tính ở bước 1. 
* Bước 3: Tách dataframe gồm năm, quốc gia, khu vực và tuổi thọ
* Bước 4: Tính toán để đưa ra các quốc gia có tuổi thọ cao nhất ở từng năm
* Bước 5: Tính toán để đưa ra các quốc gia có tuổi thọ thấp nhất ở từng năm
* Bước 6: Viết hàm để trực quan hóa các kết quả đã tính ở bước 4,5
* Bước 7: Trực quan hóa các quốc gia có tuổi thọ cao nhất, thấp nhất ở từng năm
* Bước 8: Đưa ra nhận xét

**Bước 1:** Tính trung bình tuổi thọ của **từng** quốc gia trong tập dữ liệu trong 17 năm từ năm 2000 đến 2016.

Tiến hành gom nhóm dữ liệu theo quốc gia (`country`). Sau đó, tính trung bình trên cột `life_expect`, cuối cùng sắp xếp kết quả theo thứ tự tăng dần dựa trên `life_expect`. Kết quả thu được lưu vào Series `life_exp_country`.

In [ ]:
life_exp_country =  life_exp_pro_df.groupby(['country','region']).agg(pd.Series.mean)['life_expect'].sort_values()
life_exp_country_df = life_exp_country.to_frame().reset_index()
life_exp_country_df

**Bước 2:** Trực quan hóa giá trị đã tính ở bước 1. 


In [ ]:
# Make barplot and sort bars by height
plt.figure(figsize=(16, 8))
sns.barplot(x=life_exp_country_df['country'],
            y=life_exp_country_df['life_expect'])

# Format axis' ticks
plt.xticks(fontsize=6, rotation=90)
plt.yticks(fontsize=8)

# Format labels and title
plt.xlabel('Country', size=12)
plt.ylabel("Mean Life Expectancy", size=12)
plt.title("Life Expectancy By Country",
          size=16, weight='bold')

plt.tight_layout()

**Nhận xét:** Nhật Bản là quốc gia có tuổi thọ trung bình cao nhất thế giới (2000 - 2016) với 82.75 năm, trong khi đó Sierra Leone là quốc gia có tuổi thọ trung bình thấp nhất với 46.79 năm. 

**Bước 3:** Tách dataframe gồm năm, quốc gia, khu vực và tuổi thọ và lưu vào biến `year_country_df`

In [ ]:
year_country_df = life_exp_pro_df[['year', 'country', 'region', 'life_expect']]

**Bước 4:** Tính toán để đưa ra các quốc gia có tuổi thọ cao nhất ở từng năm.

Đầu tiên tìm index của các quốc gia có tuổi thọ cao nhất theo từng năm, sau đó dựa vào index vừa tìm được để truy xuất ra các quốc gia có tuổi thọ cao nhất và lưu vào biến `countries_max`.

In [ ]:
idx_max =  year_country_df.groupby(['year'])['life_expect'].idxmax()
countries_max = year_country_df.loc[idx_max]
countries_max

**Bước 5:** Tính toán để đưa ra các quốc gia có tuổi thọ thấp nhất ở từng năm.

Đầu tiên tìm index của các quốc gia có tuổi thọ thấp nhất theo từng năm, sau đó dựa vào index vừa tìm được để truy xuất ra các quốc gia có tuổi thọ thấp nhất và lưu vào biến `countries_min`.

In [ ]:
idx_min =  year_country_df.groupby(['year'])['life_expect'].idxmin()
countries_min = year_country_df.loc[idx_min]
countries_min

**Bước 6:** Viết hàm để trực quan hóa các kết quả đã tính ở bước 4,5

In [ ]:
YEARS = tuple(set(life_exp_pro_df['year'].to_numpy()))

In [ ]:
def plot_timeline_countries(title, countries): 
    levels = np.tile([-5, 5, -3, 3, -1, 1],
                         int(np.ceil(len(YEARS)/6)))[:len(YEARS)]
    # Create figure
    fig, ax = plt.subplots(figsize=(18, 6), constrained_layout=True)
    plt.title(title, weight='bold', fontsize=20)

    # Plot stem plot; place vertical lines at each year from baseline to vertical levels
    markerline, stemline, baseline = ax.stem(YEARS, levels,\
                                             linefmt="C0-", basefmt="C9-",\
                                             use_line_collection=True)

    # Format markers
    plt.setp(markerline, mec='orange', mfc='orange', zorder=3)

    # Shift markers to the baseline by replacing y-data with zeros
    markerline.set_ydata(np.zeros(len(YEARS)))

    # Annotate stem lines
    vert = np.array(['top', 'bottom'])[(levels > 0).astype(int)]
    for yr, l, c, v in zip(YEARS, levels, countries, vert):
        ax.annotate(c, xy=(yr, l), xytext=(-3, np.sign(l)*3),
                    textcoords="offset points", va=v, ha="right", fontsize=12)

    # Format year tick labels 
    plt.setp(ax.get_xticklabels(), rotation=30, ha="right")

    # Remove y-axis and lines denoting data area boundaries (spines)
    ax.get_yaxis().set_visible(False)
    for spine in ["left", "top", "right"]:
        ax.spines[spine].set_visible(False)

    # Set padding between stemlines and x-axis
    ax.margins(y=0.1)

**Bước 7:** Trực quan hóa các quốc gia có tuổi thọ cao nhất, thấp nhất ở từng năm

In [ ]:
plot_timeline_countries("Countries with longest life expectancy over the years",\
                        countries_max['country'].to_numpy())
plot_timeline_countries("Countries with shortest life expectancy over the years",\
                        countries_min['country'].to_numpy())

**Bước 8:** Đưa ra nhận xét

* Nhật Bản là quốc gia liên tiếp có tuổi thọ cao nhất thế giới trong 17 năm khảo sát. Nguyên nhân theo nhóm tìm hiểu là do người Nhật chăm chỉ thể thao, ăn uống điều độ, nguồn gốc thực phẩm được kiểm định nghiêm ngặt,... do đó người Nhật Bản có được sức khỏe tốt, kéo dài tuổi thọ. 
* Sierra Leone, Central African Republic và Lesotho là 3 nước xuất hiện trong danh sách có tuổi thọ thấp nhất thế giới trong 17 năm khảo sát, trong đó Sierra Leone xuất hiện nhiều nhất (13 năm), Central African Republic (3 năm) và Lesotho (1 năm). Các nước này nằm cho khu vực châu Phi, có lẽ do các cuộc nội chiến, dịch vụ y tế còn yếu kém, vệ sinh an toàn thực phẩm không cao,... nên ảnh hưởng đến tuổi thọ ở các quốc gia này.

**Top 10 quốc gia có tuổi thọ trung bình cao nhất, thấp nhất?**

Các bước thực hiện: 
* **Bước 1:** Lấy ra 10 quốc gia có tuổi thọ trung bình cao nhất, 10 quốc gia có tuổi thọ trung bình thấp nhất.
* **Bước 2:** Vẽ biểu đồ tròn thể hiện sự phân bổ 10 quốc gia có tuổi thọ trung bình cao nhất / thấp nhất theo khu vực.
* **Bước 3:** Nhận xét

**Bước 1:** Lấy ra 10 quốc gia có tuổi thọ trung bình cao nhất, 10 quốc gia có tuổi thọ trung bình thấp nhất.

Với 10 quốc gia có tuổi thọ trung bình cao nhất, kết quả trả về được sắp xếp theo thứ tự giảm dần dựa trên `life_expect` và lưu vào `top_10_longest_life`. Tương tự, với 10 quốc gia có tuổi thọ trung bình thấp nhất, kết quả trả về được sắp xếp theo thứ tự tăng dần dựa trên `life_expect` và lưu vào `top_10_shortest_life`

In [ ]:
top_10_longest_life = life_exp_country_df.tail(10).sort_values(by='life_expect', ascending=False)
top_10_longest_life

In [ ]:
top_10_shortest_life = life_exp_country_df.head(10)
top_10_shortest_life

**Bước 2:** Vẽ biểu đồ tròn thể hiện sự phân bổ 10 quốc gia có tuổi thọ trung bình cao nhất / thấp nhất theo khu vực.

Đầu tiên, tính số lượng quốc gia trong các khu vực xuất hiện. Sau đó vẽ biểu đồ tròn thể hiện.

In [ ]:
num_countries_region_longest = top_10_longest_life.groupby(['region']).size()
fig = plt.figure()
ax = fig.add_axes([0,0,1,1])
ax.pie(num_countries_region_longest.values, labels = num_countries_region_longest.index,autopct='%1.2f%%')
ax.axis('equal')
ax.set_title('Percentage of the top 10 countries with longest life expect by region')
plt.show()

In [ ]:
num_countries_region_shortest = top_10_shortest_life.groupby(['region']).size()
fig = plt.figure()
ax = fig.add_axes([0,0,1,1])
ax.pie(num_countries_region_shortest.values, labels = num_countries_region_shortest.index,autopct='%1.2f%%')
ax.axis('equal')
ax.set_title('Percentage of the top 10 countries with shortest life expect by region')
plt.show()

**Bước 3:** Nhận xét
* Các quốc gia có tuổi thọ trung bình cao nhất tập trung ở khu vực châu Âu, kế đó là khu vực Tây Thái Bình Dương và khu vực châu Mỹ. 
* Các quốc gia có tuổi thọ trung bình thấp nhất tập trung chủ yếu ở khu vực châu Phi.

**Độ chênh lệch tuổi thọ năm 2016 và năm 2000 của các quốc gia như thế nào?**

Các bước thực hiện:
* **Bước 1:** Lọc ra danh sách các quốc gia (không trùng lắp) và lưu vào dataframe `change_life_exp_df`.
* **Bước 2:** Thêm vào dataframe `change_life_exp_df` cột `change` thể hiện sự thay đổi tuổi thọ giữa năm 2016 và 2000.
* **Bước 3:** Vẽ biểu đồ thể hiện và nhận xét.

**Bước 1:** Lọc ra danh sách các quốc gia (không trùng lắp) và lưu vào dataframe change_life_exp_df.

In [ ]:
change_life_exp_df = pd.DataFrame(year_country_df['country'].unique(), columns = ['country'])
change_life_exp_df

**Bước 2:** Thêm vào dataframe `change_life_exp_df` cột `change` thể hiện sự thay đổi tuổi thọ giữa năm 2016 và 2000.

In [ ]:
change_life_exp = year_country_df[year_country_df['year'] == 2016]['life_expect'].values - \
year_country_df[year_country_df['year'] == 2000]['life_expect'].values
change_life_exp_df['change'] = change_life_exp
change_life_exp_df = change_life_exp_df.sort_values(by='change')
change_life_exp_df

**Bước 3:** Vẽ biểu đồ thể hiện và nhận xét.

In [ ]:
# Make barplot and sort bars by height
plt.figure(figsize=(16, 8))
sns.barplot(x='country',
            y='change',
            data=change_life_exp_df)

# Format axis' ticks
plt.xticks(fontsize=6, rotation=90)
plt.yticks(fontsize=8)

# Format labels and title
plt.xlabel('Country', size=12)
plt.ylabel("Change Life Expectancy", size=12)
plt.title("Change in Life Expectancy Between 2000 and 2016 By Country",
          size=16, weight='bold')

plt.tight_layout()

**Nhận xét:** Nhìn chung các quốc gia có sự gia tăng tuổi thọ năm 2016 so với năm 2000, riêng quốc gia Syrian Arab Republic có sự chênh lệch mang giá trị âm, nghĩa là tuổi thọ năm 2016 giảm so với năm 2000, nguyên nhân là do nước này xảy ra cuộc nội chiến vào năm 2011 như đã trình bày ở phần khám phá dữ liệu.

**Giải thích sự biến động về mặt tuổi thọ trung bình của các quốc gia trên thế giới**

##### Về phía từng khu vực

**Tuổi thọ trung bình của các khu vực như thế nào? Đâu là khu vực có tuổi thọ trung bình cao nhất, thấp nhất?**

Các bước thực hiện:
* **Bước 1:** Gom nhóm theo năm và khu vực và tính trung bình dựa trên `life_expect`.
* **Bước 2:** Vẽ biểu độ đường thể hiện sự biến động tuổi thọ trung bình theo khu vực.
* **Bước 3:** Tính tuổi thọ trung bình của các khu vực trên thế giới.

**Bước 1:** Gom nhóm theo năm và khu vực và tính trung bình dựa trên `life_expect`.

In [ ]:
life_exp_region = life_exp_pro_df.groupby(['year', 'region']).agg(pd.Series.mean)['life_expect'].unstack(1)
life_exp_region

**Bước 2:** Vẽ biểu độ đường thể hiện sự biến động tuổi thọ trung bình theo khu vực.

In [ ]:
life_exp_region.plot.line(ylabel='Mean Life Expectancy (years)', xlabel='Year',\
                          marker='o', legend=True, figsize=(10,6))
plt.title("Mean life expectancy of regions in the world (2000-2016)", weight='bold', fontsize=14)
plt.tight_layout()
plt.show()

**Bước 3:** Tính tuổi thọ trung bình của các khu vực trên thế giới.

In [ ]:
life_exp_mean_region = life_exp_pro_df.groupby(['region']).agg(pd.Series.mean)['life_expect'].sort_values()
life_exp_mean_region

**Nhận xét về sự phân bố tuổi thọ trung bình của các khu vực trên thế giới**

* Các khu vực có tuổi thọ trung bình không đồng đều. 
* Mỗi khu vực có sự biến động khác nhau (có năm tăng, có năm giảm) nhưng nhìn chung xu hướng tuổi thọ ở các khu vực tăng. 
* Khu vực châu Âu có tuổi thọ trung bình cao nhất (75.8 năm).
* Khu vực châu Phi có tuổi thọ trung bình thấp nhất (57.7 năm).
* Nguyên nhân dẫn đến sự không đồng đều này có thể là do mỗi khu vực có sự khác nhau về kinh tế, xã hội, y tế. Những yếu tố này sẽ được khảo sát kỹ ở Câu hỏi 2.

##### Về tuổi thọ của Việt Nam

**Việt Nam đang ở vị trí nào về mặt tuổi thọ trên thế giới và trong khu vực?**

Các bước thực hiện: 
* **Bước 1:** Lấy ra dòng có giá trị `country` là `Viet Nam` từ dataframe `life_exp_country_df`
* **Bước 2:** Lấy ra các quốc gia thuộc khu vực Western Pacific
* **Bước 3:** Lấy ra dòng có giá trị `country` là `Viet Nam` từ dataframe đã tính ở bước 2.

**Bước 1:** Lấy ra dòng có giá trị country là Viet Nam từ dataframe life_exp_country_df

In [ ]:
life_exp_country_df[life_exp_country_df['country'] == 'Viet Nam']

**Nhận xét:** Việt Nam đứng ở vị trí 182 - 135 + 1 = 48 (vì số đếm tính từ 0) trong tổng số 183 quốc gia về tuổi thọ trung bình (tính từ cao xuống thấp).

**Bước 2:** Lấy ra các quốc gia thuộc khu vực Western Pacific

In [ ]:
life_exp_WP_df = life_exp_country_df[life_exp_country_df['region'] == 'Western Pacific']. \
                sort_values(by='life_expect', ascending=False).reset_index().drop(columns = ['index'])
print('So luong quoc gia trong khu vuc Tay Thai Binh Duong la: ', len(life_exp_WP_df))
life_exp_WP_df.head()

**Bước 3:** Lấy ra dòng có giá trị `country` là `Viet Nam` từ dataframe đã tính ở bước 2.

In [ ]:
life_exp_WP_df[life_exp_WP_df['country'] == 'Viet Nam']

**Nhận xét:** Việt Nam đứng ở vị trí 7 (vì số đếm tính từ 0 và dataframe đã được sort giảm dần) trong tổng số 21 quốc gia thuộc khu vực Tây Thái Bình Dương về tuổi thọ trung bình (tính từ cao xuống thấp) với tuổi thọ trung bình là 79.85 tuổi.

**Tốc độ biến động tuổi thọ ở Việt Nam qua các năm có xu hướng như thế nào?**

Các bước thực hiện:
* **Bước 1:** Lấy ra dòng có giá trị `country` là `Viet Nam` từ dataframe `life_exp_pro_df`
* **Bước 2:** Vẽ biểu đồ đường thể hiện tuổi thọ của Việt Nam qua từng năm từ 2000 đến 2016

**Bước 1:** Lấy ra dòng có giá trị `country` là `Viet Nam` từ dataframe `life_exp_pro_df`

In [ ]:
life_exp_VN = life_exp_pro_df[life_exp_pro_df['country'] == 'Viet Nam']
life_exp_VN.head()

**Bước 2:** Vẽ biểu đồ đường thể hiện tuổi thọ của Việt Nam qua từng năm từ 2000 đến 2016.

In [ ]:
life_exp_VN.plot.line(x='year', y='life_expect', ylabel='Life Expectancy (years)', xlabel='Year',\
                          marker='o', legend=True, figsize=(10,6))
plt.title("Life expectancy of Vietnam (2000-2016)", weight='bold', fontsize=14)
plt.tight_layout()
plt.show()

**Nhận xét:** Việt Nam có tuổi thọ tăng qua các năm, cụ thể là 73.34 tuổi ở năm 2000 đến 76.34 tuổi ở năm 2016. Điều này có thể được giải thích là do Việt Nam là nước đang phát triển, điều kiện về kinh tế, xã hội, giáo dục, y tế,... được đầu tư phát triển qua từng năm. Nhưng điều này cũng đồng nghĩa với việc dân số Việt Nam đang bị già hoá, do đó chính phủ cần có các biện pháp để thích ứng với việc già hóa dân số này.

---

## Nhìn lại quá trình làm đồ án

---

## Tài liệu tham khảo

---